resolwe/flow/executors/__init__.py

```diff
@@ -1,13 +1,201 @@
+                                        updates['process_error'] = process_error
+                                        updates['status'] = Data.STATUS_ERROR
+                                    elif key == 'proc.warning':
+                                        process_warning.append(val)
+                                        updates['process_warning'] = process_warning
+                                    elif key == 'proc.info':
+                                        process_info.append(val)
+                                        updates['process_info'] = process_info
+                                    elif key == 'proc.rc':
+                                        process_rc = int(val)
+                                        updates['process_rc'] = process_rc
+                                        updates['status'] = Data.STATUS_ERROR
+                                    elif key == 'proc.progress':
+                                        process_progress = int(float(val) * 100)
+                                        updates['process_progress'] = process_progress
+                                else:
+                                    dict_dot(output, key, val)
+                                    updates['output'] = output
+
+                        if updates:
+                            updates['modified'] = datetime.utcnow()
+                            Data.objects.filter(id=data_id).update(**updates)
+
+                        if process_rc > 0:
+                            log_file.close()
+                            json_file.close()
+                            os.chdir(CWD)
+                            return
+
+                        # Debug output
+                        # Not referenced in Data object
+                        json_file.write(line)
+                        json_file.flush()
+
+                except ValueError as ex:
+                    # Ignore if not JSON
+                    log_file.write(line)
+                    log_file.flush()
+
+        except MemoryError as ex:
+            logger.error(__("Out of memory: {}", ex))
+
+        except IOError as ex:
+            # TODO: if ex.errno == 28: no more free space
+            raise ex
+        finally:
+            # Store results
+            log_file.close()
+            json_file.close()
+            os.chdir(CWD)
+
+        return_code = self.end()
+
+        if process_rc < return_code:
+            process_rc = return_code
+
+        if process_rc == 0:
+            Data.objects.filter(id=data_id).update(
+                status=Data.STATUS_DONE,
+                process_progress=100,
+                finished=datetime.utcnow())
+        else:
+            Data.objects.filter(id=data_id).update(
+                status=Data.STATUS_ERROR,
+                process_progress=100,
+                process_rc=process_rc,
+                finished=datetime.utcnow())
+
+        # try:
```

Comment:
@mstajdohar this is probably commented because there was no `purge` function at the time? So I should put it back?